src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1113,6 +1116,108 @@ class FastRandomStringGenerator : public StringGenerator {
     }
 };
 
+// The following is a replacement for partial template specialization for functions, which is not
+// currently supported in C++.
+template <typename F, typename T>
+struct convert{};
+
+// Conversion used between int, long, double
+template <typename F, typename T>
+T convertToNumeric(convert<F, T>, const F& from)
+{
+    return (T)from;
+}
+
+// Special conversion for decimal
+template<typename F>
+bsoncxx::decimal128 convertToNumeric(convert<F, bsoncxx::decimal128>, const F& from) {
+    // mongocxx doesn't define nice converters from numeric types to decimal, so we convert first
+    // to string and then to decimal
+    return bsoncxx::decimal128(std::to_string(from));
+}
+
+// String conversions
+bsoncxx::decimal128 convertToNumeric(convert<std::string, bsoncxx::decimal128>,
+        const std::string& s) {
+    return bsoncxx::decimal128(s);
+}
+double convertToNumeric(convert<std::string, double>, const std::string& s) {
+    return stod(s);
+}
+int64_t convertToNumeric(convert<std::string, int64_t>, const std::string& s) {
+    return stol(s);
+}
+int32_t convertToNumeric(convert<std::string, int32_t>, const std::string& s) {
+    return stoi(s);
+}
+
+// This is the only overload we run directly, and it will route to the correct template.
+template <typename F, typename T>
+T convertToNumeric(const F& from) {
+    return convertToNumeric(convert<F, T>{}, from);
```

Comment:
This is amazing and terrible.

Reply:
Indeed. I added a brief explanation in the comment, let me know if you think that's sufficient.